lib/delete-orphans.bash

```diff
@@ -0,0 +1,226 @@
+    if [[ "$GKE_CLUSTER_NAME" != "$meta_kube_cluster" ]] ; then
+        echo "You said GKE_CLUSTER_NAME is '$GKE_CLUSTER_NAME' but Google metadata says it's '$meta_kube_cluster'"
+        echo "In case this was a mistake we wont continue."
+        exit 1
+    fi
+
+    # this is a hack to work around not having to do custom in/out of cluster kubectl commands
+    # we ignore any supplied context when running in kube, and instead make our own
+    echo "Creating a context for cluster"
+    kubectl config set-context local-cluster --namespace=default
+    KUBE_CONTEXT="local-cluster"
+}
+
+#######################################
+# Check required enviroement is setup
+#######################################
+validate() {
+    if [[ -z "$PROJECT" ]] || [[ -z "$REGION" ]] || [[ -z "$GKE_CLUSTER_NAME" ]] || [[ -z $KUBE_CONTEXT ]] ; then
+        echo "Env vars must be set"
+        echo "PROJECT:'$PROJECT'"
+        echo "REGION:'$REGION'"
+        echo "KUBE_CONTEXT:'$KUBE_CONTEXT'"
+        echo "GKE_CLUSTER_NAME:'$GKE_CLUSTER_NAME'"
+        exit 1
+    fi
+}
+
+#######################################
+# Determines if a gcloud firewall-rule is in use by the kubernetes cluster.
+# Globals:
+#   ACTIVE_IPS
+# Arguments:
+#   name of a gcloud firewall-rule resource
+# Returns:
+#   0 - the firewall-rule is in use
+#   1 - the firewall-rule is not in use
+#######################################
+valid_firewall() {
+    local id="$1"
+    local fw_json description service_name ip
+
+    fw_json=$(gcloud --project="${PROJECT}" compute firewall-rules describe "${id}" --format=json)
+    description=$(jq -r .description <<<"$fw_json")
+    service_name=$(jq -r '."kubernetes.io/service-name"' <<<"$description")
+    ip=$(jq -r '."kubernetes.io/service-ip"' <<<"$description")
+
+    echo "=> $id, IP: $ip, Service: $service_name"
+
+    if ! grep -q "$ip" <<<"$ACTIVE_IPS"; then
+        echo "  NOT in use by the kube cluster"
+        return 1
+    fi
+
+
+    # does it have a valid front end
+    echo "  IN USE by the kube cluster"
+    return 0
+}
+
+#######################################
+# Deletes all gcloud objects that represent a network load-balancer.
+# A gcloud "network load-balancer" is not a single resource but 3 resources.
+# Because kubernetes/GKE uses a similar name across all objects we are able to cleanup
+# all resources given a single name.
+# Globals:
+#   PROJECT
+#   REGION
+# Arguments:
+#   name - the gcloud name of the resources
+# Returns:
+#   None
+#######################################
+delete_gce_lb_objects() {
+    local id="$1"
+
+    # ignore failed deletions so that we can continue processing
+    set +e
+    gcloud compute "--project=${PROJECT}" -q firewall-rules   delete "k8s-fw-${id}"
+    gcloud compute "--project=${PROJECT}" -q forwarding-rules delete "${id}"        "--region=${REGION}"
+    gcloud compute "--project=${PROJECT}" -q target-pools     delete "${id}"        "--region=${REGION}"
+    wait
```

Comment:
I don't think this `wait` will do anything (basically a no-op). Did you mean to background the gcloud cmds preceeding it?

Reply:
it was there in the original. I didn't add it, but I wondered the same thing...

Reply:
seems like a simple optimization to background them all and wait, but it wont work since the API's wont allow you to delete them out of order.